File: leetcode/leetcode-39.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        li = []
        candidates.sort()
        def dfs(summ, u):
            if summ == target:
                ans.append(li[:])
                return
            for i in range(u, len(candidates)):
                if summ + candidates[i] > target:
                    return
                li.append(candidates[i])
                dfs(summ+candidates[i], i)
                li.remove(candidates[i])
        dfs(0, 0)
        return ans

if __name__ == "__main__":
    so = Solution()
    print(so.combinationSum([2,3,6,7], 7))

File: leetcode/leetcode-39.cpp
/*
 *@File Name          :leetcode-39.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/7/15 9:23
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"
using namespace std;

class Solution {
 public:
  vector<vector<int>> ans;

  void dfs(vector<int>& candidates, int target, vector<int>& t, int cur) {
    if (target == 0) {
      ans.push_back(t);
      return;
    }
    if (target < 0) return;
    if (cur == candidates.size()) return;

    for (int i = cur; i < candidates.size(); i++) {
      t.push_back(candidates[i]);
      dfs(candidates, target - candidates[i], t, i);
      t.pop_back();
    }
  }

  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<int> t;
    dfs(candidates, target, t, 0);
    return ans;
  }
};